### PR TITLE
[agent-d][issue #657] Add mailbox materialization action

### LINKED ISSUE #657: [runtime] Add deterministic system-node mailbox materialization primitive
## Summary

Empire #57 is blocked on missing Swarm support for deterministic mailbox materialization from authored events.

Empire currently emits `validation/mailbox.review_requested`, but `public.mailbox` stays empty because the only supported mailbox writer is the agent-facing `mailbox_send` tool. The approved Empire boundary requires a single deterministic root-level `mailbox-node`/system_node owner, not N prompt/tool call sites.

## Current evidence

Empire run `883eb24f-f85f-4b90-adf4-63a713a8054e` in DB `swarm_issue18_post653_20260509162413`:

- `validation/mailbox.review_requested`: 1 event
- `public.mailbox` rows for that run/entity: 0
- `opco.spinup_requested`: 0
- validation entity reached `ready_for_review`

## Platform gap

Current authoritative platform spec / runtime only allows system-node handler actions:

- `create_flow_instance`
- `record_evidence`

Runtime decoder enforces this in `internal/runtime/contracts/system_node_contract_vocab.go`, and no handler-level `mailbox_write` / mailbox materialization primitive exists. The existing `mailbox_send` implementation is an agent tool and is intentionally not the canonical owner for deterministic event materialization.

## Needed primitive

Add a fail-closed declarative system-node mailbox materialization surface, e.g. `mailbox_write`, that can:

- insert a `public.mailbox` row from a triggering event within the system-node handler transaction
- preserve source event identity (`source_event_id`)
- write deterministic row fields: item/review kind, source event, source flow/entity, payload/context, status, severity/priority, summary, idempotency key
- reject unsupported/unknown shapes at verify/boot/runtime
- avoid duplicate mailbox work on repeated delivery or fail closed with a documented invariant
- not expose prompt-owned compatibility behavior or require agents to call `mailbox_send`

## Out of scope

- Empire contract workaround using agent prompts/tools
- mailbox decision routing to `opco.spinup_requested` (Empire #60)
- OpCo BootstrapFlow design (Empire #59)
- changing existing `mailbox_send` semantics except as needed to share typed row validation, if appropriate

## Acceptance proof

- Swarm spec updated if platform semantics change
- verify rejects unsupported mailbox materialization declarations
- runtime test proves an authored system-node handler materializes exactly one mailbox row for a triggering event
- duplicate/replay proof shows idempotent behavior or fail-closed invariant
- Empire #57 can then declare the root `mailbox-node` without using agent `mailbox_send`

Blocks: yazzaoui/empire#57

### PR DESCRIPTION
**Pull request by yazzaoui**
## Summary

Closes #657.

Adds the platform-owned `mailbox_write` system-node handler action for deterministic authored event-to-`public.mailbox` materialization.

## Governing Context

- Binding issue/gate: #657 pre-audit and independent gate review, approved as first slice.
- Authoritative spec updated in `docs/specs/swarm-platform/platform/contracts/platform-spec.yaml`.
- Adjacent context: Empire #57 consumes this primitive later; Empire #60 and #59 remain split for decision routing and OpCo design.

## Semantic Migration / Ownership

- New canonical owner: Swarm handler action `mailbox_write` across platform spec, contract decoding, bootverify, and runtime pipeline action execution.
- Old non-authoritative path now invalid for this class: prompt/agent-owned `mailbox_send` as deterministic materializer for authored mailbox request events.
- Production paths checked: system-node action vocabulary, action YAML decoding, bootverify handler compliance, pipeline action registry/runner, DB mailbox row insertion, source event preservation, duplicate/replay idempotency.
- Migration completeness: complete for the Swarm primitive. Empire #57 still needs a separate product PR to declare the root mailbox node using this action.

## Proof

- `go test ./internal/runtime/contracts ./internal/runtime/bootverify ./internal/runtime/pipeline -run 'Test(SystemNodeEventHandlerDecode_.*Mailbox|Run_ReportsMailboxWrite|PipelineEngineActionRunner_MailboxWrite|PipelineEngineActionRegistry_SynthesizesSupportedBuiltinActions)'`
- `go test -run 'Test.*(Mailbox|SystemNode|Handler|Action|CreateFlowInstance|RecordEvidence)' ./internal/runtime/contracts ./internal/runtime/pipeline ./internal/store ./internal/runtime/tools ./internal/runtime/bootverify`
- `go test ./...`

### CODE REVIEW

**Comment by yazzaoui** [comment]
## Post-Implementation Proof Audit

**Semantic concept changed:** deterministic system-node materialization of authored mailbox request events into `public.mailbox`, with strict declaration shape, source-event preservation, and row-level replay idempotency.

**Chosen working failure class closed by this PR:** Swarm lacked a supported platform-owned system-node handler action for event-to-mailbox materialization. This PR adds `mailbox_write` across spec, contract decode, bootverify, runtime action execution, and tests.

**Parent failure class:** broader operator handoff lifecycle: mailbox request materialization (#657 / Empire #57), mailbox decision routing (Empire #60), and OpCo startup design (Empire #59). The parent remains open; this PR closes only the approved Swarm primitive first slice.

**Canonical owner after change:** `mailbox_write` as a platform handler action defined in `platform-spec.yaml`, loaded through `ActionSpec.Mailbox`, checked by bootverify, and executed by the pipeline action runner. `public.mailbox` remains the durable storage surface.

**Touched consumers of the owner:** system-node action vocabulary, action YAML decoder, mailbox declaration decoder, bootverify handler compliance, pipeline action registry, pipeline action runner, transaction-scoped DB insertion, platform spec docs, focused/full Go tests.

### Systematic Consumption Audit

| Consumer/seam | Status |
| --- | --- |
| Handler action vocabulary | moved to canonical owner via `IsSupportedHandlerActionID("mailbox_write")` |
| Action YAML decoding | moved to canonical owner via strict `action.mailbox` decoding |
| Unsupported declaration shapes | fail closed via strict nested mailbox/expression decoder and tests |
| Bootverify | moved to canonical owner; missing/mismatched mailbox declarations produce `handler_field_compliance` errors |
| Pipeline action registry | moved to canonical owner; synthesizes executable `mailbox_write` builtin |
| Runtime action execution | moved to canonical owner; inserts pending mailbox rows in the handler transaction |
| Row idempotency | moved to canonical owner; deterministic `item_id` from `(source_event_id, node)` and `ON CONFLICT DO NOTHING` prevents duplicate replay rows |
| `source_event_id` preservation | moved to canonical owner; runtime row stores triggering event id |
| Agent `mailbox_send` | different semantic concept; still live only as agent-facing mailbox tool, not deterministic materializer |
| Human task tools and runtime alert writers | different semantic concepts; not absorbed by this first slice |
| Empire #57 root mailbox node | not changed here; separate Empire consumer PR required |
| Empire #60/#59 | split/tracked separately |

**Old paths now invalid/non-authoritative:** prompt-owned or agent-tool `mailbox_send` materialization for authored mailbox request events; product-local `mailbox_write:` pseudo-fields outside the platform action grammar. Existing agent `mailbox_send`, human-task, and runtime alert writers still survive as separate non-closure-bearing concepts.

**Broader failure class:** operator handoff surface materialization/routing/startup. The issue was symptom-shaped around `validation/mailbox.review_requested`, but the PR implements the coherent platform primitive rather than patching that one event.

**Achieved closure level:** chosen Swarm primitive class eliminated. The parent operator handoff class remains open through Empire #57/#60/#59.

**Remaining child-slice tail:** Empire #57 must declare/verify the product `mailbox-node` using `mailbox_write`; Empire #60 must handle decision routing; Empire #59 must handle OpCo design. Estimate unchanged from pre-audit.

**Sibling contexts checked:** agent `mailbox_send`, human task mailbox tools, runtime budget/recovery alerts, mailbox readers, event receipts, action registry, action runner, bootverify, spec docs.

**Generic failing proof created:** contract decode rejects unsupported mailbox shapes; bootverify rejects missing or mismatched mailbox declarations; runtime test proves valid materialization and duplicate replay idempotency.

**Watchlist decision:** no new watchlist node. #657 is the promoted implementation issue for the primitive; broader operator-surface/watchlist coverage remains under existing #101/#95 and Empire #57/#60/#59 sequence.

**Architecture smell:** mailbox writing remains split across agent tools, human task tools, runtime alerts, and now deterministic system-node materialization. Disposition: attach to existing tracker/watchlist only. This PR intentionally avoids a generic side-effect refactor because current evidence supports a typed mailbox materializer, not a broad writer unification.

### Manifestation Coverage

| Manifestation | Status | Proof |
| --- | --- | --- |
| Authored system-node cannot declare mailbox materialization | reproduced and fixed | `TestSystemNodeEventHandlerDecode_PreservesMailboxWriteAction`; `TestPipelineEngineActionRegistry_SynthesizesSupportedBuiltinActions` |
| Unsupported mailbox declaration shape could drift into product syntax | reproduced and fixed | `TestSystemNodeEventHandlerDecode_RejectsUnknownMailboxWriteField`; `TestSystemNodeEventHandlerDecode_RejectsUnsupportedMailboxExpressionShape`; `TestRun_ReportsMailboxDeclarationOnNonMailboxAction` |
| Missing required row fields should fail before runtime ambiguity | reproduced and fixed | `TestRun_ReportsMailboxWriteMissingMailboxSpec`; `TestRun_ReportsMailboxWriteMissingRequiredFields` |
| Runtime event materializes no mailbox row | reproduced and fixed | `TestPipelineEngineActionRunner_MailboxWriteMaterializesIdempotentRow` proves pending row, source event, entity, flow, type, severity, summary, payload, producer identity |
| Replay/duplicate node execution could double-insert mailbox rows | reproduced and fixed | same runtime test executes the action twice in one transaction and proves exactly one row by deterministic item id |
| Missing runtime expression value should not create partial row | reproduced and fixed | `TestPipelineEngineActionRunner_MailboxWriteFailsClosedOnMissingRequiredExpression` |
| Empire #57 consumption | split / escalated as separate class | not claimed here; Empire must consume the Swarm primitive in a separate PR |
| Empire #60/#59 decision/OpCo path | split / escalated as separate class | explicitly out of scope and still tracked |

### Proof Run

- `go test ./internal/runtime/contracts ./internal/runtime/bootverify ./internal/runtime/pipeline -run 'Test(SystemNodeEventHandlerDecode_.*Mailbox|Run_ReportsMailboxWrite|PipelineEngineActionRunner_MailboxWrite|PipelineEngineActionRegistry_SynthesizesSupportedBuiltinActions)'`
- `go test -run 'Test.*(Mailbox|SystemNode|Handler|Action|CreateFlowInstance|RecordEvidence)' ./internal/runtime/contracts ./internal/runtime/pipeline ./internal/store ./internal/runtime/tools ./internal/runtime/bootverify`
- `go test ./...`

**Supported-surface proof:** this is a Swarm platform primitive PR, not an Empire supported-run closure. Empire #57 remains the required supported product proof after this merges.

File: docs/specs/swarm-platform/platform/platform-spec.md [hunk 1]
```diff
@@ -330,7 +330,7 @@ query (optional — pre-fetch cross-entity data into handler context)
 ### description_field
 
 - `field`: description
-- `type`: string
+- `type`: string or object
 - `purpose`: Human-readable description of what this handler does. Not executed.
```
[the remark this answers is not in the document]
Comment (yazzaoui) [reply docs/specs/swarm-platform/platform/platform-spec.md:334]: Fixed in 76c2d9d6 by restoring `description_field.type` to `string` in `platform-spec.md`, matching authoritative `platform-spec.yaml` and the runtime decoder.

File: docs/specs/swarm-platform/platform/platform-spec.md [hunk 2]
```diff
@@ -544,7 +544,7 @@ data_accumulation:
 
 - `field`: action
 - `type`: string
```
Comment (yazzaoui) [inline docs/specs/swarm-platform/platform/platform-spec.md]: `platform-spec.yaml` now declares handler `action` as `string or object`, and the new `mailbox_write` form requires an object mapping with `action.id` and `action.mailbox`. This mirrored spec still says `string`, so the published platform docs contradict the authoritative spec/runtime grammar and can lead authors or generators to reject the required mailbox form. Please regenerate/update this line to `string or object` before merge.

**Review by yazzaoui — COMMENTED** [review]
Changes needed, but GitHub would not accept a formal REQUEST_CHANGES review from the authenticated PR owner account.

The implementation and tests look structurally aligned with the approved #657 first slice: the authoritative `platform-spec.yaml` is updated, contract decode and bootverify fail closed for malformed `mailbox_write` declarations, runtime insertion uses the handler transaction context, and deterministic `item_id` gives row-level replay idempotency for the source-event/materializer tuple.

Blocking finding: `platform-spec.md` still documents the handler `action` field as `string` even though the authoritative YAML and runtime grammar now support the object form required by `mailbox_write`. That leaves the published spec internally contradictory for contract authors/tools. The earlier inline finding on `description_field.type` is obsolete on current head.

**Comment by yazzaoui** [comment]
## Review Checklist

- Guideline/spec check: changes needed. Authoritative `platform-spec.yaml` is updated, but `platform-spec.md` still says handler `action` type is `string` instead of `string or object` for the new `mailbox_write` mapping form.
- Pre-audit/gate: #657 category and class verified; approved first-slice boundary is acceptable. Empire #57/#60/#59 remain split and not claimed here.
- Inline review state: prior `description_field.type` comment is obsolete on current head; one current blocking spec-mirror finding posted.
- Runtime/code assessment: transaction-scoped insert path verified through `dbExecContext`; deterministic `item_id` provides row-level replay idempotency; no missed same-concept runtime interpreter found.
- Tests run: `go test ./internal/runtime/contracts ./internal/runtime/bootverify ./internal/runtime/pipeline -run 'Test(SystemNodeEventHandlerDecode_.*Mailbox|Run_ReportsMailboxWrite|PipelineEngineActionRunner_MailboxWrite|PipelineEngineActionRegistry_SynthesizesSupportedBuiltinActions)'`; `go test -run 'Test.*(Mailbox|SystemNode|Handler|Action|CreateFlowInstance|RecordEvidence)' ./internal/runtime/contracts ./internal/runtime/pipeline ./internal/store ./internal/runtime/tools ./internal/runtime/bootverify`; `go test ./...`.
- Risk level: medium until the spec mirror is repaired; low after that for the Swarm primitive slice.
- Closure level: Swarm `mailbox_write` primitive is otherwise close to failure-class closure, but not merge-ready until the mirrored spec contradiction is fixed.
- Merge recommendation: do not merge until `platform-spec.md` is updated/regenerated to match `platform-spec.yaml`.

File: docs/specs/swarm-platform/platform/platform-spec.md (hunk 2 above)
Comment (yazzaoui) [reply docs/specs/swarm-platform/platform/platform-spec.md]: Fixed in 9cf287b9 by updating `platform-spec.md` handler `action` type to `string or object`, matching authoritative `platform-spec.yaml` and the runtime grammar for `mailbox_write`. Also rechecked that `description_field.type` remains `string`.

**Comment by yazzaoui** [comment]
Review checklist blocker addressed in 9cf287b9: `platform-spec.md` now mirrors authoritative `platform-spec.yaml` for handler `action.type: string or object`, while `description_field.type` remains `string`. This was docs-only; `git diff --check` passed. No runtime tests rerun for the mirror-only change.

**Review by yazzaoui — COMMENTED** [review]
Re-review on head `9cf287b9`: no findings.

The prior blocking spec-mirror issue is fixed. `platform-spec.md` now mirrors authoritative `platform-spec.yaml` with handler `action` typed as `string or object`, and `description_field.type` remains `string`. The only delta from the previously reviewed/tested head is that one docs-line fix.

The approved #657 first-slice boundary still holds: this PR closes the Swarm `mailbox_write` primitive across spec, decode, bootverify, runtime action execution, and tests. It does not claim Empire #57/#60/#59 closure. I resolved the two obsolete review threads after verifying them on current head.

Merge recommendation: merge.

**Comment by yazzaoui** [comment]
## Re-Review Checklist

- Guideline/spec check: pass. `platform-spec.yaml` and `platform-spec.md` now both document handler `action` as `string or object`; `description_field.type` remains `string`.
- Prior comments: both unresolved review threads were verified obsolete on `9cf287b9` and resolved.
- Failure-class coverage: pass for the approved #657 first slice. Swarm `mailbox_write` primitive is covered; Empire #57/#60/#59 remain split and not claimed here.
- Concept-wide drift check: no missed same-concept interpreter found. Adjacent mailbox writers remain different concepts and are already tracked as non-closure-bearing architecture smell.
- Tests run on current head: `git diff --check origin/master...HEAD`; `go test ./internal/runtime/contracts ./internal/runtime/bootverify ./internal/runtime/pipeline -run 'Test(SystemNodeEventHandlerDecode_.*Mailbox|Run_ReportsMailboxWrite|PipelineEngineActionRunner_MailboxWrite|PipelineEngineActionRegistry_SynthesizesSupportedBuiltinActions)'`; `go test -run 'Test.*(Mailbox|SystemNode|Handler|Action|CreateFlowInstance|RecordEvidence)' ./internal/runtime/contracts ./internal/runtime/pipeline ./internal/store ./internal/runtime/tools ./internal/runtime/bootverify`.
- Full-suite note: `go test ./...` passed on prior head `76c2d9d6`; current head only changes one docs line in `platform-spec.md`.
- Risk level: low.
- Closure level: Swarm primitive failure class eliminated; broader operator handoff parent remains open through Empire #57/#60/#59.
- Merge recommendation: merge.
